### PR TITLE
AA-190 Composer constrainer strict mode

### PR DESCRIPTION
**Pull request by gerner-spryker**
Improvements 

- Developer(s): @gerner-spryker  

- Ticket: https://spryker.atlassian.net/browse/AA-190

- Release Group: https://release.spryker.com/release-groups/view/3798

- PR Overview:  https://release.spryker.com/release/pull-request-sdk/ComposerConstrainer/14

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   ComposerConstrainer  | patch (currently 0.x)  |                       |

-----------------------------------------

##### Change log

Improvements

* Enhanced `code:constraint:modules` console command with `--strict` argument to allow using the command in strict mode. Strict mode detects customisation/extension/configuration/dependency according to the latest Spryker conventions allowing a more accurate locking over Spryker dependencies. Also warns about customised lines per each package that will increase migration efforts upon next upgrade with Spryker packages.
* Improved `ComposerConstrainerFacade::updateConstraints()` with $isStrict parameter to enable using strict mode.
* Improved `ComposerConstrainerFacade::validateConstraints()` with $isStrict parameter to enable using strict mode.
* Introduced various amount of transfers and properties to support detailed information gathering about composer constraints for strict mode.
* Introduced `ComposerConstrainerConfig::getStrictValidationIgnoredPackages()` to allow ignoring specific packages during strict mode both for validation and update.
* Introduced `ComposerConstrainerConfig::getIsIgnoreLineCount()` to allow ignoring customization line count during evaluation that would need to be fixed otherwise.
* Enhanced `code:constraint:modules` console command with `--output-format` argument to allow different result format to be configured (currently available for strict mode only with "csv" value).

### CODE REVIEW

File: src/SprykerSdk/Zed/ComposerConstrainer/ComposerConstrainerConfig.php [hunk 1]
```diff
@@ -44,6 +44,18 @@ public function getCoreNamespaces(): array
...
 
+    /**
+     * @api
+     *
+     * @return string[]
+     */
+    public function getStrictValidationIgnoredPackages(): array
+    {
+        return [
+            'spryker/kernel$', # kernel needs different evaluation
```
Comment (dereuromark) [inline src/SprykerSdk/Zed/ComposerConstrainer/ComposerConstrainerConfig.php:55]: The closing $ is a bit weird, can we keep the string non-regex related and just use that character inside the final usage?

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+     * - Line count MUST be zero to be considered properly developed.
+     * - Expected and defined constraint lock need to match to be considered will configured.
+     *
+     * @param \Generated\Shared\Transfer\ComposerConstraintTransfer[] $composerConstraintTransfers
+     *
+     * @return \Generated\Shared\Transfer\ComposerConstraintTransfer[]
+     */
+    protected function removeCorrectPackages(array $composerConstraintTransfers): array
+    {
+        $ignoredPackages = '#(' . implode('|', $this->config->getStrictValidationIgnoredPackages()) . ')#';
```
Comment (dereuromark) [inline src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php:217]: Given the package names, I think you need ^ and $ here to make sure not to catch false positives in the long run
e.g. spryker/kernel-helper or alike in the far future (we never know).

**Comment by dereuromark** [comment]
Overall, looks good to me!

**Comment by dereuromark** [comment]
We should include the 0.* understanding (using concrete 0.x.y in those cases)

We need to also find out how we can fix up the tests
Tranfers need to be generated here for tests in order for PHPStan to work etc.

Best to merge/rebase with master
Note: Master now auto detects composer indentation (2, 4 spaces, etc) and will keep the indentation as is (to avoid noise in diff).

**Comment by dereuromark** [comment]
Can the changelogs please include what strict mode means? I think for most this is not clear from the current text.

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php (hunk 2 above)
Comment (gerner-spryker) [reply src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php:217]: i want to support magic here. On project, we actually group-ignore packages like this. This is why in Config i had the `^..$` around packages names and you asked me to remove them ;)

**Comment by gerner-spryker** [comment]
1] Can the changelogs please include what strict mode means? I think for most this is not clear from the current text.
Done

2] We should include the 0.* understanding (using concrete 0.x.y in those cases)
I won't be able to add it to this current development anymore, we are way over budget.

3] We need to also find out how we can fix up the tests. Tranfers need to be generated here for tests in order for PHPStan to work etc.
I don't get this at all :S let's discuss

4] Best to merge/rebase with master
I did that already, i saw that there was a release. But thx for the notice!

**Comment by dereuromark** [comment]
3) I did for u

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php (hunk 2 above)
Comment (dereuromark) [reply src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php:217]: OK
Well, to remove them inside the config array, not inside the actual regex, was my thinking :)

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Updater/StrictConstraintUpdater.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        if ($constraintCollectionTransfer->getComposerConstraints()->count() === 0) {
+            return new ComposerConstraintCollectionTransfer();
+        }
+
+        $composerJsonArray = $this->composerJsonReader->read();
+        foreach ($constraintCollectionTransfer->getComposerConstraints() as $composerConstraintTransfer) {
+            $moduleInfoTransfer = $composerConstraintTransfer->getModuleInfo();
+            $requireType = $moduleInfoTransfer->getIsDev() ? 'require-dev' : 'require';
+
+            $composerJsonArray[$requireType][$composerConstraintTransfer->getName()] = $moduleInfoTransfer->getExpectedConstraintLock() . $moduleInfoTransfer->getExpectedVersion();
```
Comment (dereuromark) [inline src/SprykerSdk/Zed/ComposerConstrainer/Business/Updater/StrictConstraintUpdater.php:63]: In most cases using OrFail() methods will decrease phpstan errors quite a bit, as the nullable part here causes invalid cases.

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/SprykerReflector/SprykerClassReflector.php [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        $this->fileName = $splFileInfo->getFilename();
+
+        $this->isPublicApi = (bool)preg_match('#(' . implode('|', array_merge($this->publicApiClassSuffixes, $this->publicApiInterfaceSuffixes)) . ')#', $this->fileName);
+        $this->isConfiguration = (bool)preg_match('#(' . implode('|', $this->configurationClassSuffixes) . ')#', $this->fileName);
+        $this->isFactory = (bool)preg_match('#Factory\.php#', $this->fileName);
+
+        $this->reflectionClass = new ReflectionClass($this->namespace . '\\' . $this->className);
+        $this->isInterface = $this->reflectionClass->isInterface();
+
+        $this->parentClassName = $this->reflectionClass->getParentClass() ? $this->reflectionClass->getParentClass()->getShortName() : '';
```
Comment (dereuromark) [inline src/SprykerSdk/Zed/ComposerConstrainer/Business/SprykerReflector/SprykerClassReflector.php:164]: These are PHPStan false positives
But assigning first to a local var should solve this.

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php [hunk 5]
```diff
@@ -143,8 +143,14 @@ protected function setDefinedModuleInfo(
             return $usedModuleInfo;
         }
 
-        $version = (string)($definedConstraint ? $definedConstraint->getVersion() : $featureConstraint->getVersion());
-        $isDev = (bool)($definedConstraint ? $definedConstraint->getIsDev() : $featureConstraint->getIsDev());
+        $constraint = $definedConstraint;
+        if ($constraint === null) {
+            /** @var \Generated\Shared\Transfer\ComposerConstraintTransfer $constraint */
```
Comment (dereuromark) [inline src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php:149]: This should not be necessary if the $featureConstraint cannot be null - if it was already checked for this.

**Review by dereuromark — APPROVED** [review]
I would assume that the transfer dependency should be added as >.25 to make sure the orFail functionality is available.
You can check prefer lowest first (also run tests), as here the minimum transfer pulled should properly generate the transfers.

Right now prefer lowest doesn't pull transfer at all, so I don't know

Also:
2 warnings (impossible to test minimum version here or allowed as `--majors-only`/`-m`):
 - spryker/kernel: Defined `3.33.0.0` as minimum, but is `3.56.0.0`
 - spryker/symfony: Defined `3.2.2.0` as minimum, but is `3.5.0.0`
 No biggie, it just means we cannot test the defined minimums here, so it might be safer/cleaner to set a higher one anyway.

File: src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php (hunk 5 above)
Comment (gerner-spryker) [reply src/SprykerSdk/Zed/ComposerConstrainer/Business/Validator/StrictConstraintValidator.php:149]: phpstan can't figure it out
